feat: Method that create a title with a given view 🍕

a7maad-ayman opened the pull request:
<!--
  For Work In Progress Pull Requests, please use the Draft PR feature,
  see https://github.blog/2019-02-14-introducing-draft-pull-requests/ for further details.
  
  For a timely review/response, please avoid force-pushing additional
  commits if your PR already received reviews or comments.
  
  Before submitting a Pull Request, please ensure you've done the following:
  - 📖 Read the Open Sauced Contributing Guide: https://github.com/open-sauced/open-sauced/blob/HEAD/CONTRIBUTING.md#create-a-pull-request.
  - 📖 Read the Open Sauced Code of Conduct: https://github.com/open-sauced/open-sauced/blob/HEAD/CODE_OF_CONDUCT.md.
  - 👷‍♀️ Create small PRs. In most cases, this will be possible.
  - ✅ Provide tests for your changes.
  - 📝 Use descriptive commit messages.
  - 📗 Update any related documentation and include any relevant screenshots.
-->

## What type of PR is this? (check all applicable)

- [x] 🍕 Feature

## Description
Create a function that generate header section with a provided UIView and String title

<!-- 
Please do not leave this blank 
This PR [adds/removes/fixes/replaces] the [feature/bug/etc]. 
-->
## Related Tickets & Documents
Home<>Method that create a title with a given view
<!-- 
Please use this format link issue numbers: Fixes #123
https://docs.github.com/en/free-pro-team@latest/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword 
-->
## Added tests?

- [ ] 👍 yes
- [x] 🙅 no, because they aren't needed
- [ ] 🙋 no, because I need help

## Added to documentation?

- [ ] 📜 README.md
- [ ] 📓 docs.opensauced.pizza
- [ ] 🍕 dev.to/opensauced
- [ ] 📕 storybook
- [x] 🙅 no documentation needed

## [optional] Are there any post-deployment tasks we need to perform?



## [optional] What gif best describes this PR or how it makes you feel?

a7maad-ayman committed:
feat:function that create a title with a given view

a7maad-ayman committed:
fix: Resolve linting issues

ahmdmhasn committed:
fix: Linting errors using Xcode build 👨🏼‍🔧 (#66)

* Fix linting errors

* Fix unit tests building errors

* Add colors to project

a7maad-ayman committed:
rebase the develop branch for solving linting issue

a7maad-ayman committed:
Merge branch 'develop' into feat/header-section

ahmdyasser commented on Healthy.xcodeproj/project.pbxproj:
Remove the empty lines.

ahmdyasser commented on Healthy/Classes/Modules/Dashboard/DashboardViewController.swift:
```diff
@@ -33,6 +33,54 @@ final class DashboardViewController: UIViewController {
         configureView()
     }
+
+    // MARK: - Function for header section creation
```
If you want to describe what the function does, don't use **// MARK**, instead use documentation comment. You can do so by pressing `options + cmd + /`

<img width="500" alt="image" src="https://github.com/motoon-eg/healthy/assets/42544598/e7ee2911-8032-4b92-8584-f7dade14108c">


Those info will show when you get the definition of a function like this:

<img width="500" alt="image" src="https://github.com/motoon-eg/healthy/assets/42544598/c33e5696-b837-4237-90c5-db996b719767">

Also this type of comment will help you to generate documentation for your app using DocC.

ahmdyasser commented on Healthy/Classes/Modules/Dashboard/DashboardViewController.swift:
```diff
@@ -33,6 +33,54 @@ final class DashboardViewController: UIViewController {
+        view.addSubview(stack)
+        stack.backgroundColor = .darkGray
+
+        stack.axis = .horizontal
+        stack.alignment = .center
+        stack.distribution = .fillEqually
+       stack.spacing = 10
+
+        stack.translatesAutoresizingMaskIntoConstraints = false
+
+        NSLayoutConstraint.activate([
+            stack.leadingAnchor.constraint(equalTo:
+                                            view.safeAreaLayoutGuide.leadingAnchor,
+                                           constant: 8),
+            stack.trailingAnchor.constraint(equalTo: view.safeAreaLayoutGuide.trailingAnchor,
+                                            constant: -8),
+            stack.topAnchor.constraint(equalTo: view.safeAreaLayoutGuide.topAnchor, constant: 8),
+            stack.heightAnchor.constraint(equalToConstant: view.bounds.height*0.1)
+        ])
+
+        titleLabel.translatesAutoresizingMaskIntoConstraints = false
+
+        NSLayoutConstraint.activate([
+            titleLabel.leadingAnchor.constraint(equalTo: stack.leadingAnchor, constant: 8),
+            titleLabel.trailingAnchor.constraint(equalTo: content.leadingAnchor),
+            titleLabel.topAnchor.constraint(equalTo: stack.topAnchor),
+            titleLabel.bottomAnchor.constraint(equalTo: content.topAnchor, constant: -8)
+        ])
+
+        content.translatesAutoresizingMaskIntoConstraints = false
+
+        NSLayoutConstraint.activate([
+            content.leadingAnchor.constraint(equalTo: titleLabel.trailingAnchor, constant: 10),
+            content.topAnchor.constraint(equalTo: stack.topAnchor),
+            content.trailingAnchor.constraint(equalTo: stack.trailingAnchor, constant: -8),
+            content.bottomAnchor.constraint(equalTo: content.topAnchor, constant: -8)
+        ])
+
+        return stack
+    }
```
It's a good practice to select all and indent the code whenever you finish writing code in a file, or you can download SwiftFormat to do the formatting for you.

```suggestion
    func makeHeaderSection(title: String, content: UIView) -> UIView {
        let titleLabel = UILabel()
        let stack = UIStackView(arrangedSubviews: [titleLabel, content])

        stack.translatesAutoresizingMaskIntoConstraints = false
        titleLabel.translatesAutoresizingMaskIntoConstraints = false
        content.translatesAutoresizingMaskIntoConstraints = false

        titleLabel.text = title
        view.addSubview(stack)
        stack.backgroundColor = .darkGray
        stack.axis = .horizontal
        stack.alignment = .center
        stack.distribution = .fillEqually
        stack.spacing = 10

        NSLayoutConstraint.activate([
            stack.leadingAnchor.constraint(
                equalTo: view.safeAreaLayoutGuide.leadingAnchor,
                constant: 8),
            stack.trailingAnchor.constraint(
                equalTo: view.safeAreaLayoutGuide.trailingAnchor,
                constant: -8),
            stack.topAnchor.constraint(
                equalTo: view.safeAreaLayoutGuide.topAnchor, constant: 8),
            stack.heightAnchor.constraint(equalToConstant: view.bounds.height*0.1)
        ])

        NSLayoutConstraint.activate([
            titleLabel.leadingAnchor.constraint(equalTo: stack.leadingAnchor, constant: 8),
            titleLabel.trailingAnchor.constraint(equalTo: content.leadingAnchor),
            titleLabel.topAnchor.constraint(equalTo: stack.topAnchor),
            titleLabel.bottomAnchor.constraint(equalTo: content.topAnchor, constant: -8)
        ])

        NSLayoutConstraint.activate([
            content.leadingAnchor.constraint(equalTo: titleLabel.trailingAnchor, constant: 10),
            content.topAnchor.constraint(equalTo: stack.topAnchor),
            content.trailingAnchor.constraint(equalTo: stack.trailingAnchor, constant: -8),
            content.bottomAnchor.constraint(equalTo: content.topAnchor, constant: -8)
        ])

        return stack
    }
}
```

ahmdmhasn commented on Healthy/Classes/Modules/Dashboard/DashboardViewController.swift:
```diff
@@ -33,6 +33,54 @@ final class DashboardViewController: UIViewController {
+        view.addSubview(stack)
+        stack.backgroundColor = .darkGray
+
+        stack.axis = .horizontal
+        stack.alignment = .center
+        stack.distribution = .fillEqually
+       stack.spacing = 10
+
+        stack.translatesAutoresizingMaskIntoConstraints = false
+
+        NSLayoutConstraint.activate([
+            stack.leadingAnchor.constraint(equalTo:
+                                            view.safeAreaLayoutGuide.leadingAnchor,
+                                           constant: 8),
+            stack.trailingAnchor.constraint(equalTo: view.safeAreaLayoutGuide.trailingAnchor,
+                                            constant: -8),
+            stack.topAnchor.constraint(equalTo: view.safeAreaLayoutGuide.topAnchor, constant: 8),
+            stack.heightAnchor.constraint(equalToConstant: view.bounds.height*0.1)
+        ])
+
+        titleLabel.translatesAutoresizingMaskIntoConstraints = false
+
+        NSLayoutConstraint.activate([
+            titleLabel.leadingAnchor.constraint(equalTo: stack.leadingAnchor, constant: 8),
+            titleLabel.trailingAnchor.constraint(equalTo: content.leadingAnchor),
+            titleLabel.topAnchor.constraint(equalTo: stack.topAnchor),
+            titleLabel.bottomAnchor.constraint(equalTo: content.topAnchor, constant: -8)
+        ])
+
+        content.translatesAutoresizingMaskIntoConstraints = false
+
+        NSLayoutConstraint.activate([
+            content.leadingAnchor.constraint(equalTo: titleLabel.trailingAnchor, constant: 10),
+            content.topAnchor.constraint(equalTo: stack.topAnchor),
+            content.trailingAnchor.constraint(equalTo: stack.trailingAnchor, constant: -8),
+            content.bottomAnchor.constraint(equalTo: content.topAnchor, constant: -8)
+        ])
+
+        return stack
+    }
```
This is how is think of it... No constraints are needed. 

```suggestion
// MARK: - Headers Creation

private extension DashboardViewController {
    private func makeSection(title: String, content: UIView) -> UIView {

        let titleLabel = UILabel()
        titleLabel.text = title
        titleLabel.applyStyle(TitleLabelStyle())

        let stackView = UIStackView(arrangedSubviews: [titleLabel, content])
        stackView.axis = .horizontal
        stackView.alignment = .fill
        stackView.spacing = 8.0

        return stackView
    }
}
```

This is an example usage 
```swift
    func configureView() {
        stackView.addArrangedSubview(headerView)

        // Section with header example
        stackView.addArrangedSubview(
            makeSection(
                title: "Header",
                content: UIView()
            )
        )
    }
```

ahmdmhasn replied:
Please don't forget to add one unit test case at least and add a screenshot of the result in the description.

ahmdmhasn commented on Healthy.xcodeproj/project.pbxproj:
```diff
@@ -14,8 +14,10 @@
 		022924302A33A170009290A8 /* UIColors.Generated.swift in Sources */ = {isa = PBXBuildFile; fileRef = 0229242D2A33A170009290A8 /* UIColors.Generated.swift */; };
 		0252BF192A365A6F00823339 /* CreateAccountViewModelType.swift in Sources */ = {isa = PBXBuildFile; fileRef = 0252BF182A365A6F00823339 /* CreateAccountViewModelType.swift */; };
 		025511922A3C394300295B91 /* Colors.xcassets in Resources */ = {isa = PBXBuildFile; fileRef = 20376AA82A2663F8004BF0BF /* Colors.xcassets */; };
+
```
```suggestion
```

ahmdmhasn commented on Healthy.xcodeproj/project.pbxproj:
```diff
@@ -14,8 +14,10 @@
+
 		025511962A3C3F6400295B91 /* Poppins-Regular.ttf in Resources */ = {isa = PBXBuildFile; fileRef = 025511942A3C3F6400295B91 /* Poppins-Regular.ttf */; };
 		025511972A3C3F6400295B91 /* Poppins-Bold.ttf in Resources */ = {isa = PBXBuildFile; fileRef = 025511952A3C3F6400295B91 /* Poppins-Bold.ttf */; };
+
```
```suggestion
```

ahmdmhasn commented:
@a7maad-ayman Any update?

a7maad-ayman committed:
recfator : makeSection Method

a7maad-ayman committed:
Merge branch 'develop' into feat/header-section

a7maad-ayman commented:
> @a7maad-ayman Any update?

a7maad-ayman commented:
> > @a7maad-ayman Any update?

I have made the requested changes in this PR :
 - Rebasing the `develop` branch into the local `feat/header-section` branch to incorporate the latest changes.
 - Removed unnecessary spaces and `MARK` comments .
 - Refactored the method into a new extension without any constraints 
 
 #####  Please note that I did not include Unit Tests specifically for the private method as it is not directly accessible for testing.
